Adjust public API

CendioOssman opened the pull request:
These are the changes I'd like to make before closing #924 and releasing 1.0. I'm still looking over things, but this branch should include what I've found so far.

samhed commented on core/rfb.js:
```diff
@@ -243,15 +240,13 @@ export default function RFB(defaults) {
-            return this._fail(
-                _("Must set host"));
+        if (!url) {
+            this._fail(_("Must specify URL"));
```
👎 translation change :(

CendioOssman replied:
Yes, that's unfortunate. But it's not a major issue for most users as we have a check in `ui.js` as well. So it is people using `RFB` directly that will be affected. And we haven't sorted out translations in that scenario.

CendioOssman replied:
Ideally, yes. We could also have some compromise with codes covering broad areas and an explaining text similar to how the `Error` exception is used.

Do you know how browsers handle translation in those cases? Or are they always in English?

samhed replied:
Chrome does not have translations for warnings and errors that aren't intended for end-users. But I think we can be better, it's always a bit shitty to have mixed languages if you can avoid it.

CendioOssman commented:
I'm not entirely happy with the capabilities API, so I'll have another look at that once everything else is done.

DirectXMan12 commented on core/display.js:
```diff
@@ -88,7 +81,39 @@ try {
+    // ===== PROPERTIES =====
+
+    _scale: 1.0,
+    get scale() { return this._scale; },
```
this is ES6 syntax.  Do all of our intended platforms support it?

DirectXMan12 replied:
(P.S. I'm fine with using ES6 syntax if our intended platforms support it, or even saying that we should transpile for older platforms, but it's something we need to decide on.  If we do, we can probably just switch to class syntax...)

CendioOssman replied:
MDN claims so. Haven't tested it yet.

CendioOssman commented:
I've now changed the documentation to resemble the style used by MDN. Hopefully that will make it easier for people to understand it. What do you think?

There's also a lot missing, so feel free to pitch in. :)

CendioOssman commented:
> 👎 on the moving all the public properties from separate objects to RFB -- I think we should keep those on the separate objects, but treat what's returned from the getters like an interface (think typescript style w/o actually using it):

I'm not sure I understand what you are suggesting here. Can you give a code example?

Right now I really prefer what I've suggested, because:

 - It avoid exposing our internal structure, which would otherwise force us to keep that structure
 - It makes it more obvious what is stable API and what is not. I.e. everything on the `RFB` object that isn't prefixed by `_` is fair game.
 - Having to write `rfb.get_display().scale = 0.5;` instead of just `rfb.scale = 0.5;` is just meaningless extra characters for anyone using the API as the display object is just some magical internal thing anyway.

samhed commented:
I like the MDN style API docs! 

With regards to the API for the display (and friends) towards the app; I think it ties into the discussion of whether we should treat these other modules as separately valuable or not. To me it feels like they have grown (aside from maybe keyboard) quite specific for our use case. If you agree with this it would probably be more natural to keep the API for the UI limited to the RFB module.

DirectXMan12 commented:
> I'm not sure I understand what you are suggesting here. Can you give a code example?

Without using something like TypeScript (which I'm not necessarily advocating, although it does make part of this issue much easier to deal with), it's much more just a documentation issue -- we document/explicitly mark that `scale` and a couple other things are considered stable, and everything else is not.  If someone were to provide TypeScript definition files, they'd do something like what I did above, which enforces that `let b = new RFB(); b.display.blitImage()` is a compile-time error, while `let b = new RFB(); b.display.scale = 0.5` is not.  For us, it'd just be documentation.

I'm *really* not a huge fan of piling a bunch of methods into RFB -- even if display and friends aren't usable apart from RFB, it makes sense to me to separate them.  It's easier to organize for us, and for the user, it's clearer that `rfb.display.scale` is scaling the display, as opposed to something else (the remote machine, etc).  If it's just `rfb.scale`, it's not clear what that means (IMO).

DirectXMan12 commented:
(P.S. I agree with @samhed -- the MDN-style docs are really nice)

CendioOssman commented:
> I'm really not a huge fan of piling a bunch of methods into RFB -- even if display and friends aren't usable apart from RFB, it makes sense to me to separate them.

Code wise, I agree wholeheartedly. But for the interface I don't for the reasons above. And if we need documentation, then maintaining that is similar effort to maintaining the indirection in the code. And I think it is more likely that people will check our code when figuring things out. :)

> It's easier to organize for us, and for the user, it's clearer that rfb.display.scale is scaling the display, as opposed to something else (the remote machine, etc). If it's just rfb.scale, it's not clear what that means (IMO).

I can agree with that. But that could be handles as a naming issue. So I did some changes to the names in my latest iteration of this patch set. Have a look and see what you think. :)

samhed commented:
This PR is getting huge, would it be possible to split into a few smaller ones? Would make it a lot easier to review.

CendioOssman commented:
Not easily since they build upon each other. :/

I tried to keep each commit as small as possible at least.

DirectXMan12 commented on app/ui.js:
```diff
@@ -13,7 +13,7 @@
+import { isTouchDevice } from '../core/util/browsers.js';
 import { setCapture, getPointerEvent } from '../core/util/events.js';
 import KeyTable from "../core/input/keysym.js";
 import keysyms from "../core/input/keysymdef.js";
```
make sure all of these imports are still needed.

CendioOssman replied:
`Base64` is the only one not used. But it's not used on `master` either, so feel free to remove it there.

`KeyTable` is only used by `sendCtrlAltDel()`, which I considered removing since it is just a wrapper around `sendKey()`. Does anyone of you want to keep it?

samhed replied:
I'm fine with removing sendCtrlAltDel

DirectXMan12 replied:
`Display` doesn't seem to be used either.

CendioOssman replied:
Ehm? It's instantiated in the RFB constructor. :)

CendioOssman commented on app/ui.js:
```diff
@@ -207,10 +207,9 @@ var UI = {
                               'onDisconnected': UI.disconnectFinished,
                               'onCredentials': UI.credentials,
-                              'onCapabilities': UI.updatePowerButton,
+                              'onCapabilities': function () { UI.updatePowerButton(); UI.initialResize(); },
```
Probably. I was more focused on getting the API ready than cleanups.

samhed replied:
fair enough

CendioOssman commented on app/ui.js:
```diff
@@ -277,8 +277,8 @@ var UI = {
-        UI.touchKeyboard = new Keyboard({target: document.getElementById('noVNC_keyboardinput'),
-                                         onKeyEvent: UI.keyEvent});
+        UI.touchKeyboard = new Keyboard(document.getElementById('noVNC_keyboardinput'),
+                                        {onKeyEvent: UI.keyEvent});
```
Sure. Feel free to add something to `master`. :)

CendioOssman commented on core/input/keyboard.js:
```diff
@@ -18,14 +18,14 @@ import KeyTable from "./keysym.js";
 
-export default function Keyboard(defaults) {
+export default function Keyboard(target, defaults) {
+    this._target = target || document;
```
It makes sure there's a default. I'm just keeping the existing behaviour.

samhed replied:
Yeah, but the default was null

CendioOssman replied:
Ah. Will fix.

CendioOssman commented on core/input/mouse.js:
```diff
@@ -30,7 +31,6 @@ export default function Mouse(defaults) {
 
     // Configuration attributes
     set_defaults(this, defaults, {
-        'target': document,
```
No idea. Didn't dig in to those details.

CendioOssman commented on core/rfb.js:
```diff
@@ -1478,6 +1493,8 @@ make_properties(RFB, [
     ['local_cursor', 'rw', 'bool'],         // Request locally rendered cursor
     ['shared', 'rw', 'bool'],               // Request shared mode
     ['view_only', 'rw', 'bool'],            // Disable client mouse/keyboard
+    ['scale', 'rw', 'float'],               // Display area scale factor
```
This one got fixed later. Anything you feel was overlooked?

samhed replied:
No, good!

CendioOssman commented on core/rfb.js:
```diff
@@ -1282,7 +1277,8 @@ RFB.prototype = {
                 Log.Info("XVP extensions enabled (version " + this._rfb_xvp_ver + ")");
-                this._onXvpInit(this._rfb_xvp_ver);
+                this._capabilities.power = true;
+                this._onCapabilities(this, this._capabilities)
```
I had a look at other APIs in MDN, and this seems to be a more common model.

CendioOssman commented on docs/API-internal.md:
```diff
@@ -0,0 +1,141 @@
+# 1. Internal Modules
+
+The noVNC client is a composed of several internal modules that handle
```
Inherited from the original, but I've fixed it now. :)

CendioOssman commented on core/rfb.js:
```diff
@@ -324,63 +339,55 @@ RFB.prototype = {
         }
+
+        RFB.messages.setDesktopSize(this._sock, width, height,
+                                    this._screen_id, this._screen_flags);
```
We did a cleanup in 89d2837fa8e3f0498bbac7b7194f453d88bcbc1f but apparently we missed this one.

CendioOssman replied:
`Base64` is the only one not used. But it's not used on `master` either, so feel free to remove it there.

`KeyTable` is only used by `sendCtrlAltDel()`, which I considered removing since it is just a wrapper around `sendKey()`. Does anyone of you want to keep it?

CendioOssman commented:
> Method names look better, but I'm still 👎 versus just making a couple of the methods and properties on display part of the public API.
> 

Fair enough. Which variant does @samhed and @kanaka prefer?

> Additionally, IMO, anything called from app/ui.js and app/webutil.js need to be listed as public APIs. The "app" should exist as both a standalone app and a good example of how to use all the features of core, so if someone copied over app/* as their own, they should be able to use it with the official releases of noVNC without having to worry about it using internals.

I agree that the goal should be that all features only rely on public API, but I also think that it would be very helpful if we can have `vnc.html` as a test bed for new features before we finalise the API. So I'm afraid I don't agree with the goal of being able to drop it in to any future version of `rfb.js`.

Is there something specific you noticed?

CendioOssman commented on docs/API.md:
```diff
@@ -1,178 +1,32 @@
-Display 'render_mode' option will throw an exception if an attempt is
-made to set it. The attribute mode is one of the following:
+Some attributes are read-only and cannot be changed. An exception will
+be thrown if an attempt is made to set one of these attributs. The
```
This was removed in the conversion to MDN style docs.

samhed commented:
@CendioOssman With the new names of the display functions I'm happy.

@DirectXMan12 

> Additionally, IMO, anything called from app/ui.js and app/webutil.js need to be listed as public APIs. The "app" should exist as both a standalone app and a good example of how to use all the features of core, so if someone copied over app/* as their own, they should be able to use it with the official releases of noVNC without having to worry about it using internals.

I'm not sure I follow here, you mean that app/ui.js would use internals indirectly by using for example RFB.viewportScale that in turn is using Display.scale directly?

kanaka commented:
> Fair enough. Which variant does @samhed and @kanaka prefer?

I lean towards the variant suggested by @DirectXMan12, but not strongly. My intuition is that it will be easier to maintain going forward but I don't have any concrete evidence. And since I'm not personally doing much maintenance for the foreseeable future, take that as you will.

CendioOssman commented on core/rfb.js:
```diff
@@ -230,47 +225,76 @@ export default function RFB(defaults) {
+    set viewportScale(scale) { this._display.scale = scale; },
+
+    get clipViewport() { return this._display.clipViewport; },
+    set clipViewport(viewport) { this._display.touchButton = clipViewport; },
```
Oops. Will fix. :)

CendioOssman commented:
I've updated the branch with the found problems now. I've also added a commit changing events to something more like `EventTarget`. Unfortunately it was not possible to use the browser's code for this, so something similar in JavaScript was added.

samhed commented on core/rfb.js:
```diff
@@ -512,7 +502,9 @@ RFB.prototype = {
-        this.onupdatestate(this, state, oldstate);
+        var event = document.createEvent("CustomEvent");
+        event.initCustomEvent("updatestate", false, false, { state: state });
+        this.dispatchEvent(event);
```
Can't we make a helper function for this?

CendioOssman replied:
Possibly. But it was not that much code, so I figured it might be better to keep it obvious what is happening.

samhed replied:
fair enough

DirectXMan12 commented on core/rfb.js:
```diff
@@ -512,7 +502,9 @@ RFB.prototype = {
         this._rfb_connection_state = state;
-        this.onupdatestate(this, state, oldstate);
+        var event = document.createEvent("CustomEvent");
+        event.initCustomEvent("updatestate", false, false, { state: state });
```
we should *really* be using the `CustomEvent` constructor when possible -- the `initCustomEvent` has been deprecated and removed from the standards (https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/initCustomEvent).  With our target browser support, I don't think we can use it all the time, but we can use the polyfill listed here: https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent

CendioOssman replied:
I was using the constructor to start with, but that went south once I tested on IE. :sob:

I can try the polyfill and see how that goes.

DirectXMan12 commented on core/util/eventtarget.js:
```diff
@@ -0,0 +1,45 @@
+ * See README.md for usage and integration instructions.
+ */
+
+export var EventTargetMixin = {
```
just `export default` this

CendioOssman replied:
Not sure why I didn't in the first place...

DirectXMan12 commented on core/util/eventtarget.js:
```diff
@@ -0,0 +1,45 @@
+
+   addEventListener: function(type, callback) {
+      if (!this._listeners) {
+         this._listeners = {};
```
should just use a `Map` to avoid accidental collisions with built-in properties, etc.

CendioOssman replied:
How would a `Map` improve things in that regard?

CendioOssman replied:
Ah, I see what you mean now. Will fix.

DirectXMan12 commented on core/util/eventtarget.js:
```diff
@@ -0,0 +1,45 @@
+      }
+      var stack = this._listeners[type];
+      for (var i = 0, l = stack.length; i < l; i++) {
+         if (stack[i] === callback) {
```
make this a `Set` to allow easy removal.

CendioOssman replied:
Alright, I'll give it a try.

CendioOssman commented:
Comments fixed, thanks.

I think we're at a point where we can look at merging this. I've been running the tests on most browsers, but I should probably do some manual tests as well.

We seem to be divided on the display/mouse/keyboard API issue, so unless there's a massive objection I'll refer to "he who writes the code gets to decide" to break the tie. :)

samhed commented:
LGTM!

CendioOssman commented:
Found a bunch of bugs (which I've since fixed and pushed here), but now everything seems to work fine where I can test.

I'll give you guys a few days to have a final look and then I'll merge it. :)

DirectXMan12 commented:
Whoops, looks like one of my replies didn't submitted.  I've reproduced it below.

I'll try and do a final review again soon.   Can you push this to a branch on the repo before merge?  I want to play around with submitting a PR with a couple of things that I've been thinking about, but want to keep the discussion in this repo for posterity.

> I'm not sure I follow here, you mean that app/ui.js would use internals indirectly by using for example RFB.viewportScale that in turn is using Display.scale directly?

I'm suggesting that anything we call from `app/ui.js` should be public API.  It should be theoretically possible for someone to completely reproduce `vnc.html` without using any of the internal APIs.

> I agree that the goal should be that all features only rely on public API, but I also think that it would be very helpful if we can have `vnc.html` as a test bed for new features before we finalise the API. So I'm afraid I don't agree with the goal of being able to drop it in to any future version of `rfb.js`.

That's not precisely what I was suggesting ;-).  I merely meant what I stated above -- if someone wanted to reproduce vnc.html without using internal APIs, it should be possible.  I don't think that conflicts with the goal of using `vnc.html` as a testbed.

CendioOssman commented:
> I'll try and do a final review again soon. Can you push this to a branch on the repo before merge? I want to play around with submitting a PR with a couple of things that I've been thinking about, but want to keep the discussion in this repo for posterity.

I'm not sure what you're trying to do, but I've pushed an `api` branch here now.

> That's not precisely what I was suggesting ;-). I merely meant what I stated above -- if someone wanted to reproduce vnc.html without using internal APIs, it should be possible. I don't think that conflicts with the goal of using vnc.html as a testbed.

I still don't think I understand you in that case because test bed and public api are fundamentally in conflict in my world. :)

CendioOssman commented:
@DirectXMan12, do you need some more time or can I merge this?

DirectXMan12 commented:
hey, appologies, got caught up in something at work, then got a bit sick :-/.  Anyway, I thought most of it was good to go :-).
